server/service/src/document/patient/update.rs

```diff
@@ -0,0 +1,309 @@
+use chrono::Utc;
+use repository::{DocumentRepository, EqualFilter, RepositoryError, TransactionError};
+
+use crate::{
+    document::{document_service::DocumentInsertError, raw_document::RawDocument},
+    service_provider::{ServiceContext, ServiceProvider},
+};
+
+use super::{
+    patient_doc_name, patient_schema::SchemaPatient, Patient, PatientFilter, PATIENT_TYPE,
+};
+
+#[derive(PartialEq, Debug)]
+pub enum UpdatePatientError {
+    InvalidPatientId,
+    InvalidParentId,
+    PatientExists,
+    InvalidDataSchema(Vec<String>),
+    InternalError(String),
+    DatabaseError(RepositoryError),
+}
+
+pub struct UpdatePatient {
+    pub data: serde_json::Value,
+    pub schema_id: Option<String>,
+    /// If the patient is new the parent is not set
+    pub parent: Option<String>,
+}
+
+pub fn update_patient(
+    ctx: &ServiceContext,
+    service_provider: &ServiceProvider,
+    store_id: String,
+    user_id: &str,
+    input: UpdatePatient,
+) -> Result<Patient, UpdatePatientError> {
+    let patient = ctx
+        .connection
+        .transaction_sync(|_| {
+            let patient = validate(ctx, service_provider, &store_id, &input)?;
+            let doc = generate(user_id, &patient, input);
+
+            // Updating the document will trigger an update in the patient (names) table
+            service_provider
```

Comment:
This is not clear, I would prefer service top level abstractions to look this way:

`insert_patient`
* validate
* insert document
* insert patient from document
* find and return patient (can access document through patient)

`update_patient`
* vallidate
* update document
* update patient from document
* find and return patient (can access document through patient)

Reply:
Yes should name it upsert. 

Know what you mean regarding document_updated. The only reason I haven't done it yet is that if I move the hook to the patient service I can update a patient though the low level document API without updating the names table. I am playing with the idea to restrict access to the low level document API, e.g. if the document type is Patient it can only be updated through the patient API.